flang/runtime/execute.cpp

```diff
@@ -0,0 +1,206 @@
+  if (WIFSTOPPED(status)) {
+    if (!cmdstat) {
+      terminator.Crash("stopped by signal: %d", WSTOPSIG(status));
+    } else {
+      CheckAndStoreIntToDescriptor(cmdstat, SIGNAL_ERR, terminator);
+      CopyCharsToDescriptor(*cmdmsg, "stopped by signal");
+    }
+  }
+#endif
+  return exitStatusVal;
+}
+
+void RTNAME(ExecuteCommandLine)(const Descriptor &command, bool wait,
+    const Descriptor *exitstat, const Descriptor *cmdstat,
+    const Descriptor *cmdmsg, const char *sourceFile, int line) {
+  Terminator terminator{sourceFile, line};
+  const char *newCmd{EnsureNullTerminated(
+      command.OffsetElement(), command.ElementBytes(), terminator)};
+
+  if (exitstat) {
+    RUNTIME_CHECK(terminator, IsValidIntDescriptor(exitstat));
+  }
+
+  if (cmdstat) {
+    RUNTIME_CHECK(terminator, IsValidIntDescriptor(cmdstat));
+    // Assigned 0 as specifed in standard, if error then overwrite
+    StoreIntToDescriptor(cmdstat, CMD_EXECUTED, terminator);
+  }
+
+  if (cmdmsg) {
+    RUNTIME_CHECK(terminator, IsValidCharDescriptor(cmdmsg));
+  }
+
+  if (wait) {
+    // either wait is not specified or wait is true: synchronous mode
+    int status{std::system(newCmd)};
+    int exitStatusVal{TerminationCheck(status, cmdstat, cmdmsg, terminator)};
+    // If sync, assigned processor-dependent exit status. Otherwise unchanged
+    CheckAndStoreIntToDescriptor(exitstat, exitStatusVal, terminator);
+  } else {
+// Asynchronous mode
+#ifdef _WIN32
+    STARTUPINFO si;
+    PROCESS_INFORMATION pi;
+    ZeroMemory(&si, sizeof(si));
+    si.cb = sizeof(si);
+    ZeroMemory(&pi, sizeof(pi));
+
+    // add "cmd.exe /c " to the beginning of command
+    const char *prefix{"cmd.exe /c "};
+    char *newCmdWin{(char *)AllocateMemoryOrCrash(
+        terminator, std::strlen(prefix) + std::strlen(newCmd) + 1)};
+    std::strcpy(newCmdWin, prefix);
+    std::strcat(newCmdWin, newCmd);
+
+    // Convert the char to wide char
+    const size_t sizeNeeded{mbstowcs(NULL, newCmdWin, 0) + 1};
+    wchar_t *wcmd{(wchar_t *)AllocateMemoryOrCrash(
+        terminator, sizeNeeded * sizeof(wchar_t))};
+    if (std::mbstowcs(wcmd, newCmdWin, sizeNeeded) == static_cast<size_t>(-1)) {
+      terminator.Crash("Char to wide char failed for newCmd");
+    }
+    FreeMemory((void *)newCmdWin);
+
+    if (CreateProcess(nullptr, wcmd, nullptr, nullptr, FALSE, 0, nullptr,
+            nullptr, &si, &pi)) {
+      // Close handles so it will be removed when terminated
+      CloseHandle(pi.hProcess);
+      CloseHandle(pi.hThread);
+    } else {
+      if (!cmdstat) {
+        terminator.Crash(
+            "CreateProcess failed with error code: %lu.", GetLastError());
+      } else {
+        StoreIntToDescriptor(cmdstat, (uint32_t)GetLastError(), terminator);
+        CheckAndCopyCharsToDescriptor(cmdmsg, "CreateProcess failed.");
+      }
+    }
+    FreeMemory((void *)wcmd);
+#else
+    // terminated children do not become zombies
+    signal(SIGCHLD, SIG_IGN);
```

Comment:
This actually cause a crash on async execution. Once an async `execute_command_line` was called, all `execute_command_line` after that will result in `std::system` returns a status code of -1. 
Comment out this line would be a temporary fix.

Reply:
reproducer: aarch64, linux, clang++
```cpp
#include <unistd.h>
#include <signal.h>
#include <iostream>
int execute_command_line(const char *cmd, bool isWait){
    if(isWait){
        return std::system(cmd);
    }else{
        signal(SIGCHLD, SIG_IGN);
        pid_t pid=fork();
        if(pid<0){
            return -999;
        }else if(pid==0){
            return std::system(cmd);
        }
        exit(pid);
        return -111;
    }
}

int main() {
    std::cout << execute_command_line("InvalidCommand", false) << std::endl;
    std::cout << execute_command_line("InvalidCommand", true) << std::endl;

    return 0;
}
```
console printout:
```
sh: 1: InvalidCommand: not found
-1
sh: 1: InvalidCommand: not found
-1
```
console printout when `signal(SIGCHLD, SIG_IGN);` is commented out.
```
sh: 1: InvalidCommand: not found
32512
sh: 1: InvalidCommand: not found
32512
```

Reply:
or an alternativelydo what `gfortran` is doing: do not change the value of `exitstatus`, `cmdmsg`, and set `cmdstatus` to 0, no mater what happened.

```fortran
program test_associated
  integer :: exitstatus, cmdstatus
  character(len=35) :: cmdmsg
  exitstatus=404
  cmdstatus=303
  cmdmsg="Does async execution change cmdmsg?"

  call execute_command_line ("InvalidCommand", .false., exitstatus, cmdstatus, cmdmsg)
  call sleep(1)

  print *, "Exist status was: ", exitstatus
  print *, "Command status was: ", cmdstatus
  print *, "Error message is: ", trim(cmdmsg)

  end
  ```
  console output when `isWait=false` (async)
  ```
  sh: 1: InvalidCommand: not found
 Exist status was:          404
 Command status was:            0
 Error message is: Does async execution change cmdmsg?
  ```
  
  console output when `isWait=true` (sync)
  ```
  sh: 1: InvalidCommand: not found
 Exist status was:          127
 Command status was:            3
 Error message is: Invalid command line
 ```
 
 Note that `call sleep()` is not available on flang-new, so on llvm-test-suite https://github.com/llvm/llvm-test-suite/blob/9ca97f5027150f7e507e5ab4c56f38a29fb3c696/Fortran/gfortran/regression/execute_command_line_1.f90#L16, it would fail to compile as well,

Reply:
It turns out that resetting the signal doesn't work well. put `signal(SIGCHLD, SIG_IGN)` will still cause system return -1. However, using a non-blocking `waitpid` to get the exit code of child process will clear up the zombie. (I didn't there is an non-blocking wait).
```cpp
void sigchld_handler(int signo) {
    int status;
    waitpid(-1, &status, WNOHANG); // WNOHANG means non-blocking
}
.......
signal(SIGCHLD, sigchld_handler);
pid_t pid{fork()};
if (pid < 0) {
  error;
} else if (pid == 0) {
  std::system(cmd);
}
```

Reply:
Sounds good to me :+1: 

Out of interest, what did you try to reset the handler for SIGCHLD?

Reply:
Opps, sorry I type it wrong in the comment, I mean `signal(SIGCHLD, SIG_DFL)`

Comment updated.